scrapy/utils/python.py

```diff
@@ -1,19 +1,32 @@
 """
 This module contains essential stuff that should've come with Python itself ;)
 """
+import errno
 import gc
+import inspect
 import os
 import re
-import inspect
+import sys
 import weakref
-import errno
 from functools import partial, wraps
 from itertools import chain
-import sys
 
 from scrapy.utils.decorators import deprecated
 
 
+try:
+    from dataclasses import is_dataclass, asdict as dataclass_asdict
+except ImportError:
+    def is_dataclass_instance(_):
+        return False
+
+    def dataclass_asdict(_):
+        raise ImportError("no module named 'dataclasses'")
+else:
+    def is_dataclass_instance(obj):
```

Comment:
Could you please add a bit more tests to this function?

Reply:
This is an upstream recipe, taken verbatim from the official docs: https://docs.python.org/3/library/dataclasses.html#dataclasses.is_dataclass

Reply:
Even if so, I think it is better to add tests, because the example in docs is only guaranteed to work with the Python these docs are for, while we support many Pythons, and a backport as well.